Crash: Release an un-acquired socket

### Haraka version

2.8.8 (pulled from master a few minutes ago)
### Observed behavior

```
[ERROR] [-] [core] Release an un-acquired socket. Stack: Error
    at release_client (/usr/lib/node_modules/Haraka/outbound.js:1251:69)
    at finish_processing_mail (/usr/lib/node_modules/Haraka/outbound.js:1546:9)
    at pluggableStream.<anonymous> (/usr/lib/node_modules/Haraka/outbound.js:1778:25)
    at emitOne (events.js:77:13)
    at pluggableStream.emit (events.js:169:7)
    at pluggableStream.socket.process_data (/usr/lib/node_modules/Haraka/line_socket.js:23:20)
    at pluggableStream.<anonymous> (/usr/lib/node_modules/Haraka/line_socket.js:34:48)
    at emitOne (events.js:77:13)
    at pluggableStream.emit (events.js:169:7)
    at Socket.<anonymous> (/usr/lib/node_modules/Haraka/tls_socket.js:49:14)
[CRIT] [-] [core] Error: Callback was already called.
[CRIT] [-] [core]     at /usr/lib/node_modules/Haraka/node_modules/async/lib/async.js:43:36
[CRIT] [-] [core]     at HMailItem.hmail.next_cb (/usr/lib/node_modules/Haraka/outbound.js:104:9)
[CRIT] [-] [core]     at /usr/lib/node_modules/Haraka/outbound.js:2199:15
[CRIT] [-] [core]     at FSReqWrap.oncomplete (fs.js:82:15)
```

This has reared its ugly head again.

```
root@relay:/opt/uitmail# haraka --version
Haraka.js — Version: 2.8.10
root@relay:/opt/uitmail# node --version
v4.6.2
root@relay:/opt/uitmail# 

[INFO] [-] [core] [outbound] [outbound::25:173.8.206.253:undefined:50] dispense() clients=0 available=0
[ERROR] [4FF18960-B70E-4023-B043-2CACFA74BBAB.1.1] [outbound] Ongoing connection failed to 173.8.206.253:25 : Error: write after end
[INFO] [4FF18960-B70E-4023-B043-2CACFA74BBAB.1.1] [outbound] Temp failing 1479328945801_0_13802_5983.relay for 64 seconds: Tried all MXs
[NOTICE] [4FF18960-B70E-4023-B043-2CACFA74BBAB.1.1] [outbound] delivered file=1479328945801_0_13802_5983.relay domain="happykidsdentistry.com" host="staff.happykidsdentistry.com" ip=173.8.206.253 port=25 mode=SMTP tls=Y auth=N response="" delay=67.609 fails=1 rcpts=1/0/0
[ERROR] [-] [core] Release an un-acquired socket. Stack: Error
    at release_client (/usr/lib/node_modules/Haraka/outbound.js:1298:69)
    at finish_processing_mail (/usr/lib/node_modules/Haraka/outbound.js:1597:9)
    at pluggableStream.<anonymous> (/usr/lib/node_modules/Haraka/outbound.js:1812:25)
    at emitOne (events.js:77:13)
    at pluggableStream.emit (events.js:169:7)
    at pluggableStream.socket.process_data (/usr/lib/node_modules/Haraka/line_socket.js:23:20)
    at pluggableStream.<anonymous> (/usr/lib/node_modules/Haraka/line_socket.js:34:48)
    at emitOne (events.js:77:13)
    at pluggableStream.emit (events.js:169:7)
    at TLSSocket.<anonymous> (/usr/lib/node_modules/Haraka/tls_socket.js:47:14)
[CRIT] [-] [core] Error: Callback was already called.
[CRIT] [-] [core]     at /usr/lib/node_modules/Haraka/node_modules/async/dist/async.js:844:36
[CRIT] [-] [core]     at HMailItem.hmail.next_cb (/usr/lib/node_modules/Haraka/outbound.js:106:9)
[CRIT] [-] [core]     at /usr/lib/node_modules/Haraka/outbound.js:2233:15
[CRIT] [-] [core]     at FSReqWrap.oncomplete (fs.js:82:15)
[NOTICE] [-] [core] Shutting down
root@relay:/opt/uitmail# 

```

Hi, I'm using haraka as an outbound relay in front of an Exchange 2003.
It works perfect almost all the time, and i love for its simplicity and flexibility but I had this crash now.

I'm using two further instances in the inbound chain where are no problems.

node: 4.2.6
haraka: 2.8.11


> 2016-12-22T12:22:04.642Z [INFO] [-] [core] [outbound] [outbound::25:94.199.88.70:undefined:0] dispense() clients=0 available=0
2016-12-22T12:22:25.684Z [PROTOCOL] [5F880B78-9903-4D8F-88EA-80597AE94048.1.1] [outbound] S: 250 2.0.0 Ok: queued as 3tkrH21w4BzQkCM\r\n
2016-12-22T12:22:25.684Z [PROTOCOL] [5F880B78-9903-4D8F-88EA-80597AE94048.1.1] [outbound] C: RSET
2016-12-22T12:22:25.686Z [ERROR] [5F880B78-9903-4D8F-88EA-80597AE94048.1.1] [outbound] Ongoing connection failed to 94.199.88.70:25 : Error: write after end
2016-12-22T12:22:25.686Z [DEBUG] [-] [core] [outbound] release_client: 94.199.88.70:25 to undefined
2016-12-22T12:22:25.686Z [DEBUG] [-] [core] [outbound] destroying pool entry for 94.199.88.70:25
2016-12-22T12:22:25.686Z [INFO] [5F880B78-9903-4D8F-88EA-80597AE94048.1.1] [outbound] Attempting to deliver to: 217.111.120.3:25 (0) (1)
2016-12-22T12:22:25.687Z [INFO] [-] [core] [outbound] [outbound::25:217.111.120.3:undefined:0] dispense() clients=1 available=0
2016-12-22T12:22:25.687Z [DEBUG] [-] [core] [outbound] [outbound::25:217.111.120.3:undefined:0] createResource() - creating obj - count=1 min=0 max=10
2016-12-22T12:22:25.688Z [DEBUG] [-] [core] [outbound] host=217.111.120.3 port=25 pool_timeout=0 created
2016-12-22T12:22:25.688Z [NOTICE] [5F880B78-9903-4D8F-88EA-80597AE94048.1.1] [outbound] delivered file=1482409293799_0_1152_8571.mailgateway domain="h-f.at" host="h-f.at.mailin3.rmx.de" ip=94.199.88.70 port=25 mode=SMTP tls=Y auth=N response="" delay=51.889 fails=0 rcpts=1/0/0
2016-12-22T12:22:25.688Z [DEBUG] [5F880B78-9903-4D8F-88EA-80597AE94048.1.1] [outbound] running delivered hooks
2016-12-22T12:22:25.689Z [DEBUG] [-] [core] [outbound] release_client: 94.199.88.70:25 to undefined
2016-12-22T12:22:25.694Z [ERROR] [-] [core] Release an un-acquired socket. Stack: Error
    at release_client (/usr/local/lib/node_modules/Haraka/outbound.js:1294:69)
    at finish_processing_mail (/usr/local/lib/node_modules/Haraka/outbound.js:1593:9)
    at pluggableStream.<anonymous> (/usr/local/lib/node_modules/Haraka/outbound.js:1808:25)
    at emitOne (events.js:77:13)
    at pluggableStream.emit (events.js:169:7)
    at pluggableStream.socket.process_data (/usr/local/lib/node_modules/Haraka/line_socket.js:24:20)
    at pluggableStream.<anonymous> (/usr/local/lib/node_modules/Haraka/line_socket.js:35:48)
    at emitOne (events.js:77:13)
    at pluggableStream.emit (events.js:169:7)
    at TLSSocket.<anonymous> (/usr/local/lib/node_modules/Haraka/tls_socket.js:56:14)

> 2016-12-22T12:22:26.019Z [ERROR] [5F880B78-9903-4D8F-88EA-80597AE94048.1.1] [outbound] Reading from the data stream failed: Error: ENOENT: no such file or directory, open '/var/spool/haraka/outbound/queue/1482409293799_0_1152_8571.mailgateway'
2016-12-22T12:22:56.020Z [INFO] [-] [core] [outbound] [outbound::25:217.111.120.3:undefined:0] dispense() clients=0 available=0
2016-12-22T12:22:56.041Z [ERROR] [5F880B78-9903-4D8F-88EA-80597AE94048.1.1] [outbound] Remote end 217.111.120.3:25 closed connection while we were processing mail. Trying next MX.
2016-12-22T12:22:56.041Z [DEBUG] [-] [core] [outbound] release_client: 217.111.120.3:25 to undefined
2016-12-22T12:22:56.042Z [DEBUG] [-] [core] [outbound] destroying pool entry for 217.111.120.3:25
2016-12-22T12:22:56.042Z [INFO] [5F880B78-9903-4D8F-88EA-80597AE94048.1.1] [outbound] Attempting to deliver to: 62.245.148.18:25 (0) (1)
2016-12-22T12:22:56.042Z [INFO] [-] [core] [outbound] [outbound::25:62.245.148.18:undefined:0] dispense() clients=1 available=0
2016-12-22T12:22:56.042Z [DEBUG] [-] [core] [outbound] [outbound::25:62.245.148.18:undefined:0] createResource() - creating obj - count=1 min=0 max=10
2016-12-22T12:22:56.042Z [DEBUG] [-] [core] [outbound] host=62.245.148.18 port=25 pool_timeout=0 created
2016-12-22T12:22:56.391Z [ERROR] [5F880B78-9903-4D8F-88EA-80597AE94048.1.1] [outbound] Reading from the data stream failed: Error: ENOENT: no such file or directory, open '/var/spool/haraka/outbound/queue/1482409293799_0_1152_8571.mailgateway'
2016-12-22T12:23:26.391Z [INFO] [-] [core] [outbound] [outbound::25:62.245.148.18:undefined:0] dispense() clients=0 available=0
2016-12-22T12:23:26.415Z [ERROR] [5F880B78-9903-4D8F-88EA-80597AE94048.1.1] [outbound] Remote end 62.245.148.18:25 closed connection while we were processing mail. Trying next MX.
2016-12-22T12:23:26.415Z [DEBUG] [-] [core] [outbound] release_client: 62.245.148.18:25 to undefined
2016-12-22T12:23:26.415Z [DEBUG] [-] [core] [outbound] destroying pool entry for 62.245.148.18:25
2016-12-22T12:23:26.415Z [INFO] [5F880B78-9903-4D8F-88EA-80597AE94048.1.1] [outbound] Attempting to deliver to: 94.199.90.79:25 (0) (1)
2016-12-22T12:23:26.416Z [INFO] [-] [core] [outbound] [outbound::25:94.199.90.79:undefined:0] dispense() clients=1 available=0
2016-12-22T12:23:26.416Z [DEBUG] [-] [core] [outbound] [outbound::25:94.199.90.79:undefined:0] createResource() - creating obj - count=1 min=0 max=10
2016-12-22T12:23:26.418Z [DEBUG] [-] [core] [outbound] host=94.199.90.79 port=25 pool_timeout=0 created
2016-12-22T12:23:26.777Z [ERROR] [5F880B78-9903-4D8F-88EA-80597AE94048.1.1] [outbound] Reading from the data stream failed: Error: ENOENT: no such file or directory, open '/var/spool/haraka/outbound/queue/1482409293799_0_1152_8571.mailgateway'
2016-12-22T12:23:56.777Z [INFO] [-] [core] [outbound] [outbound::25:94.199.90.79:undefined:0] dispense() clients=0 available=0
2016-12-22T12:23:56.797Z [ERROR] [5F880B78-9903-4D8F-88EA-80597AE94048.1.1] [outbound] Remote end 94.199.90.79:25 closed connection while we were processing mail. Trying next MX.
2016-12-22T12:23:56.797Z [DEBUG] [-] [core] [outbound] release_client: 94.199.90.79:25 to undefined
2016-12-22T12:23:56.797Z [DEBUG] [-] [core] [outbound] destroying pool entry for 94.199.90.79:25
2016-12-22T12:23:56.798Z [INFO] [5F880B78-9903-4D8F-88EA-80597AE94048.1.1] [outbound] Looking up A records for: h-f.at.mailin2.rmx.de
2016-12-22T12:23:56.816Z [INFO] [5F880B78-9903-4D8F-88EA-80597AE94048.1.1] [outbound] Attempting to deliver to: 94.199.90.77:25 (0) (1)
2016-12-22T12:23:56.816Z [INFO] [-] [core] [outbound] [outbound::25:94.199.90.77:undefined:0] dispense() clients=1 available=0
2016-12-22T12:23:56.816Z [DEBUG] [-] [core] [outbound] [outbound::25:94.199.90.77:undefined:0] createResource() - creating obj - count=1 min=0 max=10
2016-12-22T12:23:56.816Z [DEBUG] [-] [core] [outbound] host=94.199.90.77 port=25 pool_timeout=0 created
2016-12-22T12:23:57.178Z [ERROR] [5F880B78-9903-4D8F-88EA-80597AE94048.1.1] [outbound] Reading from the data stream failed: Error: ENOENT: no such file or directory, open '/var/spool/haraka/outbound/queue/1482409293799_0_1152_8571.mailgateway'
2016-12-22T12:24:27.178Z [INFO] [-] [core] [outbound] [outbound::25:94.199.90.77:undefined:0] dispense() clients=0 available=0
2016-12-22T12:24:27.197Z [ERROR] [5F880B78-9903-4D8F-88EA-80597AE94048.1.1] [outbound] Remote end 94.199.90.77:25 closed connection while we were processing mail. Trying next MX.
2016-12-22T12:24:27.197Z [DEBUG] [-] [core] [outbound] release_client: 94.199.90.77:25 to undefined
2016-12-22T12:24:27.197Z [DEBUG] [-] [core] [outbound] destroying pool entry for 94.199.90.77:25
2016-12-22T12:24:27.197Z [INFO] [5F880B78-9903-4D8F-88EA-80597AE94048.1.1] [outbound] Attempting to deliver to: 217.111.120.2:25 (0) (1)
2016-12-22T12:24:27.198Z [INFO] [-] [core] [outbound] [outbound::25:217.111.120.2:undefined:0] dispense() clients=1 available=0
2016-12-22T12:24:27.198Z [DEBUG] [-] [core] [outbound] [outbound::25:217.111.120.2:undefined:0] createResource() - creating obj - count=1 min=0 max=10
2016-12-22T12:24:27.198Z [DEBUG] [-] [core] [outbound] host=217.111.120.2 port=25 pool_timeout=0 created
2016-12-22T12:24:27.532Z [ERROR] [5F880B78-9903-4D8F-88EA-80597AE94048.1.1] [outbound] Reading from the data stream failed: Error: ENOENT: no such file or directory, open '/var/spool/haraka/outbound/queue/1482409293799_0_1152_8571.mailgateway'
2016-12-22T12:24:57.532Z [INFO] [-] [core] [outbound] [outbound::25:217.111.120.2:undefined:0] dispense() clients=0 available=0
2016-12-22T12:24:57.553Z [ERROR] [5F880B78-9903-4D8F-88EA-80597AE94048.1.1] [outbound] Remote end 217.111.120.2:25 closed connection while we were processing mail. Trying next MX.
2016-12-22T12:24:57.553Z [DEBUG] [-] [core] [outbound] release_client: 217.111.120.2:25 to undefined
2016-12-22T12:24:57.553Z [DEBUG] [-] [core] [outbound] destroying pool entry for 217.111.120.2:25
2016-12-22T12:24:57.553Z [INFO] [5F880B78-9903-4D8F-88EA-80597AE94048.1.1] [outbound] Attempting to deliver to: 62.245.148.17:25 (0) (1)
2016-12-22T12:24:57.553Z [INFO] [-] [core] [outbound] [outbound::25:62.245.148.17:undefined:0] dispense() clients=1 available=0
2016-12-22T12:24:57.554Z [DEBUG] [-] [core] [outbound] [outbound::25:62.245.148.17:undefined:0] createResource() - creating obj - count=1 min=0 max=10
2016-12-22T12:24:57.554Z [DEBUG] [-] [core] [outbound] host=62.245.148.17 port=25 pool_timeout=0 created
2016-12-22T12:24:57.889Z [ERROR] [5F880B78-9903-4D8F-88EA-80597AE94048.1.1] [outbound] Reading from the data stream failed: Error: ENOENT: no such file or directory, open '/var/spool/haraka/outbound/queue/1482409293799_0_1152_8571.mailgateway'
2016-12-22T12:25:27.889Z [INFO] [-] [core] [outbound] [outbound::25:62.245.148.17:undefined:0] dispense() clients=0 available=0
2016-12-22T12:25:27.910Z [ERROR] [5F880B78-9903-4D8F-88EA-80597AE94048.1.1] [outbound] Remote end 62.245.148.17:25 closed connection while we were processing mail. Trying next MX.
2016-12-22T12:25:27.910Z [DEBUG] [-] [core] [outbound] release_client: 62.245.148.17:25 to undefined
2016-12-22T12:25:27.911Z [DEBUG] [-] [core] [outbound] destroying pool entry for 62.245.148.17:25
2016-12-22T12:25:27.911Z [INFO] [5F880B78-9903-4D8F-88EA-80597AE94048.1.1] [outbound] Attempting to deliver to: 94.199.88.69:25 (0) (1)
2016-12-22T12:25:27.911Z [INFO] [-] [core] [outbound] [outbound::25:94.199.88.69:undefined:0] dispense() clients=1 available=0
2016-12-22T12:25:27.911Z [DEBUG] [-] [core] [outbound] [outbound::25:94.199.88.69:undefined:0] createResource() - creating obj - count=1 min=0 max=10
2016-12-22T12:25:27.911Z [DEBUG] [-] [core] [outbound] host=94.199.88.69 port=25 pool_timeout=0 created
2016-12-22T12:25:28.241Z [ERROR] [5F880B78-9903-4D8F-88EA-80597AE94048.1.1] [outbound] Reading from the data stream failed: Error: ENOENT: no such file or directory, open '/var/spool/haraka/outbound/queue/1482409293799_0_1152_8571.mailgateway'
2016-12-22T12:25:58.242Z [INFO] [-] [core] [outbound] [outbound::25:94.199.88.69:undefined:0] dispense() clients=0 available=0
2016-12-22T12:25:58.261Z [ERROR] [5F880B78-9903-4D8F-88EA-80597AE94048.1.1] [outbound] Remote end 94.199.88.69:25 closed connection while we were processing mail. Trying next MX.
2016-12-22T12:25:58.261Z [DEBUG] [-] [core] [outbound] release_client: 94.199.88.69:25 to undefined
2016-12-22T12:25:58.261Z [DEBUG] [-] [core] [outbound] destroying pool entry for 94.199.88.69:25
2016-12-22T12:25:58.262Z [INFO] [5F880B78-9903-4D8F-88EA-80597AE94048.1.1] [outbound] Looking up A records for: h-f.at.mailin4.rmx.de
2016-12-22T12:25:58.281Z [INFO] [5F880B78-9903-4D8F-88EA-80597AE94048.1.1] [outbound] Attempting to deliver to: 62.245.148.19:25 (0) (1)
2016-12-22T12:25:58.281Z [INFO] [-] [core] [outbound] [outbound::25:62.245.148.19:undefined:0] dispense() clients=1 available=0
2016-12-22T12:25:58.281Z [DEBUG] [-] [core] [outbound] [outbound::25:62.245.148.19:undefined:0] createResource() - creating obj - count=1 min=0 max=10
2016-12-22T12:25:58.282Z [DEBUG] [-] [core] [outbound] host=62.245.148.19 port=25 pool_timeout=0 created
2016-12-22T12:25:58.712Z [ERROR] [5F880B78-9903-4D8F-88EA-80597AE94048.1.1] [outbound] Reading from the data stream failed: Error: ENOENT: no such file or directory, open '/var/spool/haraka/outbound/queue/1482409293799_0_1152_8571.mailgateway'
2016-12-22T12:26:28.712Z [INFO] [-] [core] [outbound] [outbound::25:62.245.148.19:undefined:0] dispense() clients=0 available=0
2016-12-22T12:26:28.733Z [ERROR] [5F880B78-9903-4D8F-88EA-80597AE94048.1.1] [outbound] Remote end 62.245.148.19:25 closed connection while we were processing mail. Trying next MX.
2016-12-22T12:26:28.733Z [DEBUG] [-] [core] [outbound] release_client: 62.245.148.19:25 to undefined
2016-12-22T12:26:28.733Z [DEBUG] [-] [core] [outbound] destroying pool entry for 62.245.148.19:25
2016-12-22T12:26:28.733Z [INFO] [5F880B78-9903-4D8F-88EA-80597AE94048.1.1] [outbound] Attempting to deliver to: 217.111.120.4:25 (0) (1)
2016-12-22T12:26:28.733Z [INFO] [-] [core] [outbound] [outbound::25:217.111.120.4:undefined:0] dispense() clients=1 available=0
2016-12-22T12:26:28.733Z [DEBUG] [-] [core] [outbound] [outbound::25:217.111.120.4:undefined:0] createResource() - creating obj - count=1 min=0 max=10
2016-12-22T12:26:28.734Z [DEBUG] [-] [core] [outbound] host=217.111.120.4 port=25 pool_timeout=0 created
2016-12-22T12:26:29.095Z [ERROR] [5F880B78-9903-4D8F-88EA-80597AE94048.1.1] [outbound] Reading from the data stream failed: Error: ENOENT: no such file or directory, open '/var/spool/haraka/outbound/queue/1482409293799_0_1152_8571.mailgateway'
2016-12-22T12:26:59.096Z [INFO] [-] [core] [outbound] [outbound::25:217.111.120.4:undefined:0] dispense() clients=0 available=0
2016-12-22T12:26:59.116Z [ERROR] [5F880B78-9903-4D8F-88EA-80597AE94048.1.1] [outbound] Remote end 217.111.120.4:25 closed connection while we were processing mail. Trying next MX.
2016-12-22T12:26:59.116Z [DEBUG] [-] [core] [outbound] release_client: 217.111.120.4:25 to undefined
2016-12-22T12:26:59.116Z [DEBUG] [-] [core] [outbound] destroying pool entry for 217.111.120.4:25
2016-12-22T12:26:59.117Z [INFO] [5F880B78-9903-4D8F-88EA-80597AE94048.1.1] [outbound] Attempting to deliver to: 82.135.18.227:25 (0) (1)
2016-12-22T12:26:59.117Z [INFO] [-] [core] [outbound] [outbound::25:82.135.18.227:undefined:0] dispense() clients=1 available=0
2016-12-22T12:26:59.117Z [DEBUG] [-] [core] [outbound] [outbound::25:82.135.18.227:undefined:0] createResource() - creating obj - count=1 min=0 max=10
2016-12-22T12:26:59.117Z [DEBUG] [-] [core] [outbound] host=82.135.18.227 port=25 pool_timeout=0 created
2016-12-22T12:26:59.454Z [ERROR] [5F880B78-9903-4D8F-88EA-80597AE94048.1.1] [outbound] Reading from the data stream failed: Error: ENOENT: no such file or directory, open '/var/spool/haraka/outbound/queue/1482409293799_0_1152_8571.mailgateway'
2016-12-22T12:27:29.454Z [INFO] [-] [core] [outbound] [outbound::25:82.135.18.227:undefined:0] dispense() clients=0 available=0
2016-12-22T12:27:29.475Z [ERROR] [5F880B78-9903-4D8F-88EA-80597AE94048.1.1] [outbound] Remote end 82.135.18.227:25 closed connection while we were processing mail. Trying next MX.
2016-12-22T12:27:29.475Z [DEBUG] [-] [core] [outbound] release_client: 82.135.18.227:25 to undefined
2016-12-22T12:27:29.475Z [DEBUG] [-] [core] [outbound] destroying pool entry for 82.135.18.227:25
2016-12-22T12:27:29.476Z [INFO] [5F880B78-9903-4D8F-88EA-80597AE94048.1.1] [outbound] Looking up A records for: h-f.at.mailin1.rmx.de
2016-12-22T12:27:29.495Z [INFO] [5F880B78-9903-4D8F-88EA-80597AE94048.1.1] [outbound] Attempting to deliver to: 217.111.120.6:25 (0) (1)
2016-12-22T12:27:29.495Z [INFO] [-] [core] [outbound] [outbound::25:217.111.120.6:undefined:0] dispense() clients=1 available=0
2016-12-22T12:27:29.495Z [DEBUG] [-] [core] [outbound] [outbound::25:217.111.120.6:undefined:0] createResource() - creating obj - count=1 min=0 max=10
2016-12-22T12:27:29.496Z [DEBUG] [-] [core] [outbound] host=217.111.120.6 port=25 pool_timeout=0 created
2016-12-22T12:27:29.833Z [ERROR] [5F880B78-9903-4D8F-88EA-80597AE94048.1.1] [outbound] Reading from the data stream failed: Error: ENOENT: no such file or directory, open '/var/spool/haraka/outbound/queue/1482409293799_0_1152_8571.mailgateway'
2016-12-22T12:27:59.835Z [INFO] [-] [core] [outbound] [outbound::25:217.111.120.6:undefined:0] dispense() clients=0 available=0
2016-12-22T12:27:59.854Z [ERROR] [5F880B78-9903-4D8F-88EA-80597AE94048.1.1] [outbound] Remote end 217.111.120.6:25 closed connection while we were processing mail. Trying next MX.
2016-12-22T12:27:59.855Z [DEBUG] [-] [core] [outbound] release_client: 217.111.120.6:25 to undefined
2016-12-22T12:27:59.856Z [DEBUG] [-] [core] [outbound] destroying pool entry for 217.111.120.6:25
2016-12-22T12:27:59.856Z [INFO] [5F880B78-9903-4D8F-88EA-80597AE94048.1.1] [outbound] Attempting to deliver to: 94.199.90.76:25 (0) (1)
2016-12-22T12:27:59.856Z [INFO] [-] [core] [outbound] [outbound::25:94.199.90.76:undefined:0] dispense() clients=1 available=0
2016-12-22T12:27:59.856Z [DEBUG] [-] [core] [outbound] [outbound::25:94.199.90.76:undefined:0] createResource() - creating obj - count=1 min=0 max=10
2016-12-22T12:27:59.858Z [DEBUG] [-] [core] [outbound] host=94.199.90.76 port=25 pool_timeout=0 created
2016-12-22T12:28:00.184Z [ERROR] [5F880B78-9903-4D8F-88EA-80597AE94048.1.1] [outbound] Reading from the data stream failed: Error: ENOENT: no such file or directory, open '/var/spool/haraka/outbound/queue/1482409293799_0_1152_8571.mailgateway'
2016-12-22T12:28:30.184Z [INFO] [-] [core] [outbound] [outbound::25:94.199.90.76:undefined:0] dispense() clients=0 available=0
2016-12-22T12:28:30.204Z [ERROR] [5F880B78-9903-4D8F-88EA-80597AE94048.1.1] [outbound] Remote end 94.199.90.76:25 closed connection while we were processing mail. Trying next MX.
2016-12-22T12:28:30.204Z [DEBUG] [-] [core] [outbound] release_client: 94.199.90.76:25 to undefined
2016-12-22T12:28:30.204Z [DEBUG] [-] [core] [outbound] destroying pool entry for 94.199.90.76:25
2016-12-22T12:28:30.205Z [INFO] [5F880B78-9903-4D8F-88EA-80597AE94048.1.1] [outbound] Attempting to deliver to: 217.111.120.1:25 (0) (1)
2016-12-22T12:28:30.205Z [INFO] [-] [core] [outbound] [outbound::25:217.111.120.1:undefined:0] dispense() clients=1 available=0
2016-12-22T12:28:30.205Z [DEBUG] [-] [core] [outbound] [outbound::25:217.111.120.1:undefined:0] createResource() - creating obj - count=1 min=0 max=10
2016-12-22T12:28:30.205Z [DEBUG] [-] [core] [outbound] host=217.111.120.1 port=25 pool_timeout=0 created
2016-12-22T12:28:30.600Z [ERROR] [5F880B78-9903-4D8F-88EA-80597AE94048.1.1] [outbound] Reading from the data stream failed: Error: ENOENT: no such file or directory, open '/var/spool/haraka/outbound/queue/1482409293799_0_1152_8571.mailgateway'
2016-12-22T12:29:00.600Z [INFO] [-] [core] [outbound] [outbound::25:217.111.120.1:undefined:0] dispense() clients=0 available=0
2016-12-22T12:29:00.620Z [ERROR] [5F880B78-9903-4D8F-88EA-80597AE94048.1.1] [outbound] Remote end 217.111.120.1:25 closed connection while we were processing mail. Trying next MX.
2016-12-22T12:29:00.620Z [DEBUG] [-] [core] [outbound] release_client: 217.111.120.1:25 to undefined
2016-12-22T12:29:00.620Z [DEBUG] [-] [core] [outbound] destroying pool entry for 217.111.120.1:25
2016-12-22T12:29:00.620Z [INFO] [5F880B78-9903-4D8F-88EA-80597AE94048.1.1] [outbound] Attempting to deliver to: 217.111.120.5:25 (0) (1)
2016-12-22T12:29:00.621Z [INFO] [-] [core] [outbound] [outbound::25:217.111.120.5:undefined:0] dispense() clients=1 available=0
2016-12-22T12:29:00.621Z [DEBUG] [-] [core] [outbound] [outbound::25:217.111.120.5:undefined:0] createResource() - creating obj - count=1 min=0 max=10
2016-12-22T12:29:00.621Z [DEBUG] [-] [core] [outbound] host=217.111.120.5 port=25 pool_timeout=0 created
2016-12-22T12:29:00.979Z [ERROR] [5F880B78-9903-4D8F-88EA-80597AE94048.1.1] [outbound] Reading from the data stream failed: Error: ENOENT: no such file or directory, open '/var/spool/haraka/outbound/queue/1482409293799_0_1152_8571.mailgateway'
2016-12-22T12:29:30.979Z [INFO] [-] [core] [outbound] [outbound::25:217.111.120.5:undefined:0] dispense() clients=0 available=0
2016-12-22T12:29:30.999Z [ERROR] [5F880B78-9903-4D8F-88EA-80597AE94048.1.1] [outbound] Remote end 217.111.120.5:25 closed connection while we were processing mail. Trying next MX.
2016-12-22T12:29:30.999Z [DEBUG] [-] [core] [outbound] release_client: 217.111.120.5:25 to undefined
2016-12-22T12:29:31.000Z [DEBUG] [-] [core] [outbound] destroying pool entry for 217.111.120.5:25
2016-12-22T12:29:31.001Z [DEBUG] [-] [core] Temp fail for: Tried all MXs
2016-12-22T12:29:31.001Z [DEBUG] [5F880B78-9903-4D8F-88EA-80597AE94048.1.1] [outbound] running deferred hooks
2016-12-22T12:29:31.052Z [INFO] [5F880B78-9903-4D8F-88EA-80597AE94048.1.1] [outbound] Temp failing 1482409293799_0_1152_8571.mailgateway for 64 seconds: Tried all MXs
2016-12-22T12:29:31.052Z [INFO] [5F880B78-9903-4D8F-88EA-80597AE94048.1.1] [outbound] bouncing mail: Error re-queueing email: Error: ENOENT: no such file or directory, rename '/var/spool/haraka/outbound/queue/1482409293799_0_1152_8571.mailgateway' -> '/var/spool/haraka/outbound/queue/1482409835052_1_1152_8571.mailgateway'
2016-12-22T12:29:31.052Z [DEBUG] [5F880B78-9903-4D8F-88EA-80597AE94048.1.1] [outbound] running bounce hooks
2016-12-22T12:29:31.102Z [ERROR] [5F880B78-9903-4D8F-88EA-80597AE94048.1.1] [outbound] Double bounce: Error populating bounce message: Error: ENOENT: no such file or directory, open '/var/spool/haraka/outbound/queue/1482409293799_0_1152_8571.mailgateway'
2016-12-22T12:29:31.105Z [CRIT] [-] [core] Error: Callback was already called.
2016-12-22T12:29:31.105Z [CRIT] [-] [core]     at /usr/local/lib/node_modules/Haraka/node_modules/async/dist/async.js:837:32
2016-12-22T12:29:31.106Z [CRIT] [-] [core]     at HMailItem.hmail.next_cb (/usr/local/lib/node_modules/Haraka/outbound.js:101:9)
2016-12-22T12:29:31.106Z [CRIT] [-] [core]     at HMailItem.double_bounce (/usr/local/lib/node_modules/Haraka/outbound.js:2138:10)
2016-12-22T12:29:31.106Z [CRIT] [-] [core]     at /usr/local/lib/node_modules/Haraka/outbound.js:2122:25
2016-12-22T12:29:31.106Z [CRIT] [-] [core]     at ReadStream.<anonymous> (/usr/local/lib/node_modules/Haraka/outbound.js:1890:13)
2016-12-22T12:29:31.106Z [CRIT] [-] [core]     at emitOne (events.js:77:13)
2016-12-22T12:29:31.106Z [CRIT] [-] [core]     at ReadStream.emit (events.js:169:7)
2016-12-22T12:29:31.106Z [CRIT] [-] [core]     at fs.js:1682:12
2016-12-22T12:29:31.106Z [CRIT] [-] [core]     at FSReqWrap.oncomplete (fs.js:82:15)
2016-12-22T12:29:31.107Z [NOTICE] [-] [core] Shutting down

Hi,
had this issue again 2 times today,

before everything crashes it starts always with the ERROR Ongoing connection failed to . . "write after end"


```
2017-01-09T11:02:06.244Z [PROTOCOL] [CA14CEA2-7D77-4081-A988-B6AEA46F1597.1.1] [outbound] S: 250 2.6.0 <0A542B54AA449246AFB91297853166D0070E65CB@domain> [InternalId=675619] Queued mail for delivery\r\n
2017-01-09T11:02:06.244Z [PROTOCOL] [CA14CEA2-7D77-4081-A988-B6AEA46F1597.1.1] [outbound] C: RSET
2017-01-09T11:02:06.246Z [ERROR] [CA14CEA2-7D77-4081-A988-B6AEA46F1597.1.1] [outbound] Ongoing connection failed to 93.82.177.214:25 : Error: write after end
```

Are you on 2.8.12? We upgraded node-generic-pool to see if this fixed
things.

On Mon, Jan 9, 2017 at 8:10 AM, shurli <notifications@github.com> wrote:

> Hi,
> had this issue again 2 times today,
>
> before everything crashes it starts always with the ERROR Ongoing
> connection failed to . . "write after end"
>
> 2017-01-09T11:02:06.244Z [PROTOCOL] [CA14CEA2-7D77-4081-A988-B6AEA46F1597.1.1] [outbound] S: 250 2.6.0 <0A542B54AA449246AFB91297853166D0070E65CB@domain> [InternalId=675619] Queued mail for delivery\r\n
> 2017-01-09T11:02:06.244Z [PROTOCOL] [CA14CEA2-7D77-4081-A988-B6AEA46F1597.1.1] [outbound] C: RSET
> 2017-01-09T11:02:06.246Z [ERROR] [CA14CEA2-7D77-4081-A988-B6AEA46F1597.1.1] [outbound] Ongoing connection failed to 93.82.177.214:25 : Error: write after end
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/haraka/Haraka/issues/1611#issuecomment-271280532>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAobYyHKuTBlMeY-I9SHDCNiMpRVmxjvks5rQjHHgaJpZM4KG7n1>
> .
>

--- a/outbound.js
+++ b/outbound.js
@@ -1246,7 +1246,7 @@ function get_pool (port, host, local_addr, is_unix_socket, connect_timeout, pool
         var pool = generic_pool.Pool({
             name: name,
             create: _create_socket,
-            validate: function(socket) {
+            validate: function (socket) {
                 return socket.writable;
             },
             destroy: function (socket) {
@@ -1288,7 +1288,7 @@ function get_client (port, host, local_addr, is_unix_socket, callback) {
     if (cfg.pool_concurrency_max == 0) {
         return _create_socket(port, host, local_addr, is_unix_socket, cfg.connect_timeout, cfg.pool_timeout, callback);
     }
-    
+
     var pool = get_pool(port, host, local_addr, is_unix_socket, cfg.connect_timeout, cfg.pool_timeout, cfg.pool_concurrency_max);
     if (pool.waitingClientsCount() >= cfg.pool_concurrency_max) {
         return callback("Too many waiting clients for pool", null);
@@ -1428,7 +1428,7 @@ HMailItem.prototype.try_deliver_host_on_socket = function (mx, host, port, socke
             return self.try_deliver_host(mx);
         }
     });
-    
+
     var fin_sent = false;
     socket.once('end', function () {
         fin_sent = true;

--- a/outbound.js
+++ b/outbound.js
@@ -1210,6 +1210,29 @@ var cram_md5_response = function (username, password, challenge) {
     return utils.base64(username + ' ' + digest);
 }
 
+function _create_socket (port, host, local_addr, is_unix_socket, connect_timeout, pool_timeout, callback) {
+    var socket = is_unix_socket ? sock.connect({path: host}) :
+        sock.connect({port: port, host: host, localAddress: local_addr});
+    socket.setTimeout(connect_timeout * 1000);
+    logger.logdebug('[outbound] host=' +
+        host + ' port=' + port + ' pool_timeout=' + pool_timeout + ' created');
+    socket.once('connect', function () {
+        socket.removeAllListeners('error'); // these get added after callback
+        callback(null, socket);
+    });
+    socket.once('error', function (err) {
+        socket.end();
+        if (server.notes.pool[name]) {
+            delete server.notes.pool[name];
+        }
+        callback("Outbound connection error: " + err, null);
+    });
+    socket.once('timeout', function () {
+        socket.end();
+        callback("Outbound connection timed out to " + host + ":" + port, null);
+    });
+}
+
 // Separate pools are kept for each set of server attributes.
 function get_pool (port, host, local_addr, is_unix_socket, connect_timeout, pool_timeout, max) {
     port = port || 25;
@@ -1222,29 +1245,8 @@ function get_pool (port, host, local_addr, is_unix_socket, connect_timeout, pool
     if (!server.notes.pool[name]) {
         var pool = generic_pool.Pool({
             name: name,
-            create: function (callback) {
-                var socket = is_unix_socket ? sock.connect({path: host}) :
-                    sock.connect({port: port, host: host, localAddress: local_addr});
-                socket.setTimeout(connect_timeout * 1000);
-                logger.logdebug('[outbound] host=' +
-                    host + ' port=' + port + ' pool_timeout=' + pool_timeout + ' created');
-                socket.once('connect', function () {
-                    socket.removeAllListeners('error'); // these get added after callback
-                    callback(null, socket);
-                });
-                socket.once('error', function (err) {
-                    socket.end();
-                    if (server.notes.pool[name]) {
-                        delete server.notes.pool[name];
-                    }
-                    callback("Outbound connection error: " + err, null);
-                });
-                socket.once('timeout', function () {
-                    socket.end();
-                    callback("Outbound connection timed out to " + host + ":" + port, null);
-                });
-            },
-            validate: function (socket) {
+            create: _create_socket,
+            validate: function(socket) {
                 return socket.writable;
             },
             destroy: function (socket) {
@@ -1283,6 +1285,10 @@ function get_pool (port, host, local_addr, is_unix_socket, connect_timeout, pool
 
 // Get a socket for the given attributes.
 function get_client (port, host, local_addr, is_unix_socket, callback) {
+    if (cfg.pool_concurrency_max == 0) {
+        return _create_socket(port, host, local_addr, is_unix_socket, cfg.connect_timeout, cfg.pool_timeout, callback);
+    }
+    
     var pool = get_pool(port, host, local_addr, is_unix_socket, cfg.connect_timeout, cfg.pool_timeout, cfg.pool_concurrency_max);
     if (pool.waitingClientsCount() >= cfg.pool_concurrency_max) {
         return callback("Too many waiting clients for pool", null);
@@ -1421,6 +1427,11 @@ HMailItem.prototype.try_deliver_host_on_socket = function (mx, host, port, socke
             release_client(socket, port, host, mx.bind, true);
             return self.try_deliver_host(mx);
         }
+    });
+    
+    var fin_sent = false;
+    socket.once('end', function () {
+        fin_sent = true;
     });
 
     var command = mx.using_lmtp ? 'connect_lmtp' : 'connect';
@@ -1597,7 +1608,7 @@ HMailItem.prototype.try_deliver_host_on_socket = function (mx, host, port, socke
         else {
             self.discard();
         }
-        release_client(socket, port, host, mx.bind);
+        release_client(socket, port, host, mx.bind, fin_sent);
     };
 
     socket.on('line', function (line) {
@@ -1787,8 +1798,8 @@ HMailItem.prototype.try_deliver_host_on_socket = function (mx, host, port, socke
                                 send_command('DATA');
                             }
                             else {
-                                send_command('RSET');
                                 finish_processing_mail(false);
+                                send_command('RSET');
                             }
                         }
                         else {

--- a/outbound.js
+++ b/outbound.js
@@ -648,7 +648,7 @@ exports.send_trans_email = function (transaction, next) {
 
 exports.process_delivery = function (ok_paths, todo, hmails, cb) {
     var self = this;
-    this.loginfo("Processing domain: " + todo.domain);
+    logger.loginfo("[outbound] Processing domain: " + todo.domain);
     var fname = _fname();
     var tmp_path = path.join(queue_dir, platformDOT + fname);
     var ws = new FsyncWriteStream(tmp_path, { flags: WRITE_EXCL });
@@ -656,7 +656,7 @@ exports.process_delivery = function (ok_paths, todo, hmails, cb) {
         var dest_path = path.join(queue_dir, fname);
         fs.rename(tmp_path, dest_path, function (err) {
             if (err) {
-                self.logerror("Unable to rename tmp file!: " + err);
+                logger.logerror("[outbound] Unable to rename tmp file!: " + err);
                 fs.unlink(tmp_path, function () {});
                 cb("Queue error");
             }
@@ -668,7 +668,7 @@ exports.process_delivery = function (ok_paths, todo, hmails, cb) {
         });
     });
     ws.on('error', function (err) {
-        self.logerror("Unable to write queue file (" + fname + "): " + err);
+        logger.logerror("[outbound] Unable to write queue file (" + fname + "): " + err);
         ws.destroy();
         fs.unlink(tmp_path, function () {});
         cb("Queueing failed");
@@ -716,7 +716,7 @@ exports.split_to_new_recipients = function (hmail, recipients, response, cb) {
     var tmp_path = path.join(queue_dir, platformDOT + fname);
     var ws = new FsyncWriteStream(tmp_path, { flags: WRITE_EXCL });
     var err_handler = function (err, location) {
-        self.logerror("Error while splitting to new recipients (" + location + "): " + err);
+        logger.logerror("[outbound] Error while splitting to new recipients (" + location + "): " + err);
         hmail.todo.rcpt_to.forEach(function (rcpt) {
             hmail.extend_rcpt_with_dsn(rcpt, DSN.sys_unspecified("Error splitting to new recipients: " + err));
         });
@@ -756,7 +756,7 @@ exports.split_to_new_recipients = function (hmail, recipients, response, cb) {
     };
 
     ws.on('error', function (err) {
-        self.logerror("Unable to write queue file (" + fname + "): " + err);
+        logger.logerror("[outbound] Unable to write queue file (" + fname + "): " + err);
         ws.destroy();
         hmail.todo.rcpt_to.forEach(function (rcpt) {
             hmail.extend_rcpt_with_dsn(rcpt, DSN.sys_unspecified("Error re-queueing some recipients: " + err));

--- a/outbound.js
+++ b/outbound.js
@@ -859,30 +859,6 @@ exports.HMailItem = HMailItem;
 
 logger.add_log_methods(HMailItem.prototype, "outbound");
 
-// populate log functions - so we can use hooks
-for (var key in logger) {
-    if (key.match(/^log\w/)) {
-        exports[key] = (function (key2) {
-            return function () {
-                var args = ["[outbound] "];
-                for (var i=0, l=arguments.length; i<l; i++) {
-                    args.push(arguments[i]);
-                }
-                logger[key2].apply(logger, args);
-            };
-        })(key);
-        HMailItem.prototype[key] = (function (key2) {
-            return function () {
-                var args = [ this ];
-                for (var i=0, l=arguments.length; i<l; i++) {
-                    args.push(arguments[i]);
-                }
-                logger[key2].apply(logger, args);
-            };
-        })(key);
-    }
-}
-
 HMailItem.prototype.data_stream = function () {
     return fs.createReadStream(this.path, {start: this.data_start, end: this.file_size});
 };

--- a/outbound.js
+++ b/outbound.js
@@ -857,6 +857,8 @@ function HMailItem (filename, filePath, notes) {
 util.inherits(HMailItem, events.EventEmitter);
 exports.HMailItem = HMailItem;
 
+logger.add_log_methods(HMailItem.prototype, "outbound");
+
 // populate log functions - so we can use hooks
 for (var key in logger) {
     if (key.match(/^log\w/)) {
@@ -1222,6 +1224,7 @@ function _create_socket (port, host, local_addr, is_unix_socket, connect_timeout
     });
     socket.once('error', function (err) {
         socket.end();
+        var name = 'outbound::' + port + ':' + host + ':' + local_addr + ':' + pool_timeout;
         if (server.notes.pool[name]) {
             delete server.notes.pool[name];
         }

--- a/outbound.js
+++ b/outbound.js
@@ -1264,7 +1264,7 @@ function get_pool (port, host, local_addr, is_unix_socket, connect_timeout, pool
                     logger.logwarn("[outbound] Socket got an error while shutting down: " + err);
                 });
                 if (!socket.writable) return;
-                logger.logprotocol("C: QUIT");
+                logger.logprotocol("[outbound] C: QUIT");
                 socket.write("QUIT\r\n");
                 socket.end(); // half close
                 socket.once('line', function (line) {

--- a/outbound.js
+++ b/outbound.js
@@ -1282,6 +1282,10 @@ function get_client (port, host, local_addr, is_unix_socket, callback) {
 function release_client (socket, port, host, local_addr, error) {
     logger.logdebug("[outbound] release_client: " + host + ":" + port + " to " + local_addr);
 
+    if (cfg.pool_concurrency_max == 0) {
+        return sockend();
+    }
+
     if (!socket.__acquired) {
         logger.logerror("Release an un-acquired socket. Stack: " + (new Error()).stack);
         return;

--- a/tests/outbound.js
+++ b/tests/outbound.js
@@ -37,6 +37,17 @@ exports.outbound = {
             test.deepEqual(lines.join('\r\n'), result);
         });
         test.done();
+    },
+    'log_methods added': function (test) {
+        var logger = require('../logger');
+        test.expect(Object.keys(logger.levels).length);
+
+        var HMailItem = require('../outbound').HMailItem;
+
+        Object.keys(logger.levels).forEach(function (level) {
+            test.ok(HMailItem.prototype['log' + level.toLowerCase()], "Log method for level: " + level);
+        });
+        test.done();
     }
 };
 

--- a/tests/outbound_protocol/basic_outbound_trial_test.js
+++ b/tests/outbound_protocol/basic_outbound_trial_test.js
@@ -56,6 +56,7 @@ function runBasicSmtpConversation (hmail) {
 function _runBasicSmtpConversation (hmail) {
     var mock_socket = mock_sock.connect('testhost', 'testport');
     mock_socket.writable = true;
+    mock_socket.__acquired = true;
 
     // The playbook
     // from remote: This line is to be sent (from an mocked remote SMTP) to haraka outbound. This is done in this test.

